src/lib.rs

```diff
@@ -271,8 +271,8 @@ pub async fn get_build_output(
                 build_string: recipe
                     .build()
                     .string()
-                    .expect("Shouldn't be unset, needs major refactoring, for handling this better")
-                    .to_owned(),
+                    .map(ToString::to_string)
+                    .unwrap_or_else(|| format!("{}_{}", hash, recipe.build().number())),
```

Comment:
Maybe the logic that the default build string is hash joined with build number could be put into a function that is also used by the `build_string` function

Reply:
You are right. I changed the implementation to use a special type to make sure this cant go wrong in the future. 

Actually while doing this refactor I found a couple of different cases where this DID go wrong previously! Nice catch @Hofer-Julian

Reply:
Typestate pattern ❤️